internal/server/singleprocess/state/trigger.go

```diff
@@ -0,0 +1,383 @@
+package state
+
+import (
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/go-memdb"
+	bolt "go.etcd.io/bbolt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+var triggerBucket = []byte("trigger")
+
+func init() {
+	dbBuckets = append(dbBuckets, triggerBucket)
+	dbIndexers = append(dbIndexers, (*State).triggerIndexInit)
+	schemas = append(schemas, triggerIndexSchema)
+}
+
+// TriggerPut creates or updates the given Trigger.
+func (s *State) TriggerPut(t *pb.Trigger) error {
+	memTxn := s.inmem.Txn(true)
+	defer memTxn.Abort()
+
+	err := s.db.Update(func(dbTxn *bolt.Tx) error {
+		if t.Project == nil {
+			return status.Error(codes.FailedPrecondition, "A Project is required to create a trigger")
+		}
+
+		if t.Id != "" {
+			var err error
+			_, err = s.triggerGet(dbTxn, memTxn, &pb.Ref_Trigger{Id: t.Id})
+			if err != nil {
```

Comment:
Am I reading this right? This prevents upserts, only updates. So if you specify an ID, you must be updated an existing trigger? If that's true, that doesn't match our other `Put` functions, and you may want to split this into a `Create` and `Update` separately.

Reply:
I think that's a mistake on my part. We should be ok to allow upserts at the state API level here.

Reply:
Fixed this up to be a proper Put.